### PR TITLE
ARROW-10100: [C++][Python][Dataset] Add ParquetFileFragment::Subset method

### PR DESCRIPTION
**Pull request by jorisvandenbossche**

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@bkietz this is what I had in mind for https://issues.apache.org/jira/browse/ARROW-10100. Thoughts?

(only POC, if we add it, need to add proper tests etc)

File: cpp/src/arrow/dataset/file_parquet.cc [hunk 1]
```diff
@@ -575,6 +575,39 @@ Result<FragmentVector> ParquetFileFragment::SplitByRowGroup(
...
+  ARROW_ASSIGN_OR_RAISE(auto new_fragment,
+                        parquet_format_.MakeFragment(source_, partition_expression(),
+                                                     row_groups, physical_schema_));
+  return new_fragment;
+}
+
+Result<std::shared_ptr<Fragment>> ParquetFileFragment::Subset(
+    const std::vector<int> row_group_ids) {
+  std::vector<RowGroupInfo> row_groups;
+
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/dataset/file_parquet.cc:592]: Here I would still need to check that all row_group_ids exist in the fragment

**Comment by jorisvandenbossche** [comment]
@bkietz so I was now thinking it is actually fine to return a fragment with 0 row groups (it can then be the responsibility of the user to check for this, if they want to filter out the empty fragments out of their list of fragments). But, the problem is that an empty row group vector currently already means "all row groups" ... 
Would it be fine to change empty vector to nullptr to indicate "all row groups" ?

File: cpp/src/arrow/dataset/file_parquet.h [hunk 2]
```diff
@@ -232,6 +232,10 @@ class ARROW_DS_EXPORT ParquetFileFragment : public FileFragment {
   /// \brief Ensure attached statistics are complete and the physical schema is cached.
   Status EnsureCompleteMetadata(parquet::arrow::FileReader* reader = NULLPTR);
 
+  /// \brief Return a filtered subset of the ParquetFileFragment.
+  Result<std::shared_ptr<Fragment>> Subset(const std::shared_ptr<Expression>& predicate);
+  Result<std::shared_ptr<Fragment>> Subset(const std::vector<int> row_group_ids);
```
Comment (bkietz) [inline cpp/src/arrow/dataset/file_parquet.h:237]:
```suggestion
  Result<std::shared_ptr<ParquetFileFragment>> Subset(const std::shared_ptr<Expression>& predicate);
  Result<std::shared_ptr<ParquetFileFragment>> Subset(const std::vector<int> row_group_ids);
```

File: cpp/src/arrow/dataset/file_parquet.cc [hunk 3]
```diff
@@ -631,6 +631,39 @@ Result<FragmentVector> ParquetFileFragment::SplitByRowGroup(
...
 
+Result<std::shared_ptr<Fragment>> ParquetFileFragment::Subset(
+    const std::shared_ptr<Expression>& predicate) {
+  RETURN_NOT_OK(EnsureCompleteMetadata());
+  ARROW_ASSIGN_OR_RAISE(auto row_groups, FilterRowGroups(*predicate));
+
+  ARROW_ASSIGN_OR_RAISE(auto new_fragment,
+                        parquet_format_.MakeFragment(source_, partition_expression(),
+                                                     row_groups, physical_schema_));
+  return new_fragment;
```
Comment (bkietz) [inline cpp/src/arrow/dataset/file_parquet.cc:642]:
```suggestion
  return checked_pointer_cast<ParquetFileFragment>(std::move(new_fragment));
```

**Review by bkietz — COMMENTED** [review]
Being more explicit about "lack of subselection" vs "empty subselection" sounds fine to me. I'll push a patch

**Comment by jorisvandenbossche** [comment]
@bkietz Thanks for the update! I further expanded the tests somewhat for the empty case.

**Review by kszucs — APPROVED** [review]
LGTM, build error is unrelated.

**Pull request closed without merging**